File: src/Commands/CallCommand.cs
namespace Statesman.Commands
{
    public class CallCommand : Command
    {
        public enum CallType
        {
            None,
            Normal,
            Super,
            Global
        }

        public const string kIdGoto = "goto";
        public const string kIdCall = "call";
        public const string kIdCallGlobal = "callglob";
        public const string kIdSuper = "super";

        public CallType CallerType { get; private set; }
        private readonly string _functionName;

        public CallCommand(CallType callType, string functionName)
        {
            if (callType == CallType.None ||
                (callType != CallType.Super &&
                 string.IsNullOrWhiteSpace(functionName)))
            {
                throw new ArgumentException("Function name cannot be empty.");
            }
            _functionName = functionName;
            CallerType = callType;
        }

        public new static Command FromText(string commandId, string[] arguments)
        {
            CallType callType = CallType.None;
            string functionName = "";

            if (arguments.Length == 1 &&
                commandId.Equals(kIdSuper, StringComparison.InvariantCultureIgnoreCase))
            {
                callType = CallType.Super;
            }
            else if (arguments.Length == 2)
            {
                if (commandId.Equals(kIdGoto, StringComparison.InvariantCultureIgnoreCase) ||
                    commandId.Equals(kIdCall, StringComparison.InvariantCultureIgnoreCase))
                {
                    callType = CallType.Normal;
                }
                else if (commandId.Equals(kIdCallGlobal, StringComparison.InvariantCultureIgnoreCase))
                {
                    callType = CallType.Global;
                }
                functionName = arguments[1];
            }

            if (callType == CallType.None)
            {
                return null;
            }

            return new CallCommand(callType, functionName);
        }

        public override void Execute()
        {
            if (CallerType == CallType.Super)
            {
                return;
            }

            Content.Script.Functions.TryGetValue(_functionName, out Function function);

            // Local functions override global functions if we're a normal call.
            if (Interpreter.Scene != null && CallerType == CallType.Normal)
            {
                if (Interpreter.Scene.Functions.TryGetValue(_functionName, out Function localFunction))
                {
                    localFunction?.Execute();
                    return;
                }
            }

            function?.Execute();
        }
    }
}

File: src/statesman/commands/CallCommand.java
package statesman.commands;

import statesman.Content;
import statesman.Interpreter;

public class CallCommand extends Command {

    public enum CallType {
        NORMAL, SUPER, GLOBAL
    }
    
    public static final String ID_GOTO = "goto";
    public static final String ID_CALL = "call";
    public static final String ID_CALL_GLOB = "callglob";
    public static final String ID_SUPER = "super";

    private CallType _callType;
    private String _functionName;

    private CallCommand() {
        _functionName = "";
        _callType = null;
    }

    public CallCommand(CallType callType, String functionName) {
        if (callType == null) {
            throw new IllegalArgumentException();
        }
        if (callType != CallType.SUPER
                && (functionName == null || functionName.isBlank())) {
            throw new IllegalArgumentException();
        }
        _functionName = functionName;
        _callType = callType;
    }

    private static Command _defaultInstance;

    public static Command getDefault() {
        if (_defaultInstance == null) {
            _defaultInstance = new CallCommand();
        }
        return _defaultInstance;
    }

    public CallType getCallType() {
        return _callType;
    }

    @Override
    public Command fromText(String commandId, String[] arguments) {
        CallType callType = null;
        String functionName = "";

        if (arguments.length == 1 && commandId.equalsIgnoreCase(ID_SUPER)) {
            callType = CallType.SUPER;
        } else if (arguments.length == 2) {
            if (commandId.equalsIgnoreCase(ID_CALL)
                    || commandId.equalsIgnoreCase(ID_GOTO)) {
                callType = CallType.NORMAL;
            } else if (commandId.equalsIgnoreCase(ID_CALL_GLOB)) {
                callType = CallType.GLOBAL;
            }
            functionName = arguments[1];
        }
        
        if (callType == null) {
            return null;
        }

        return new CallCommand(callType, functionName);
    }

    @Override
    public void execute() {
        // Return early because we're handled by the containing function.
        if (_callType == CallType.SUPER) {
            return;
        }

        Function globalFunction = Content.getScript().getFunctions()
                .get(_functionName);

        // Local functions override global functions if we're a normal call.
        if (Interpreter.getScene() != null && _callType == CallType.NORMAL) {
            Function localFunction = Interpreter
                    .getScene()
                    .getFunctions()
                    .get(_functionName);
            if (localFunction != null) {
                localFunction.execute();
                return;
            }
        }

        if (globalFunction != null) {
            globalFunction.execute();
        }
    }

}
